File: leetcode-go/1049_lastStoneWeightII.go
package leetcode_go

func lastStoneWeightII(stones []int) int {
	dp := make([]int, 15001)
	sum := 0
	for _, v := range stones {
		sum += v
	}
	target := sum / 2
	for _, stone := range stones {
		for i := target; i >= stone; i-- {
			dp[i] = max(dp[i], dp[i-stone]+stone)
		}
	}
	return sum - 2*dp[target]
}

File: src/dp/_1049_last_stone_weight_ii.rs
struct Solution;

impl Solution {
    pub fn last_stone_weight_ii(stones: Vec<i32>) -> i32 {
        let sum: i32 = stones.iter().sum();
        let mut dp = vec![0; 15001];
        let target = sum / 2;
        for stone in stones {
            for i in (stone..=target).rev() {
                dp[i as usize] = dp[i as usize].max(dp[(i - stone) as usize] + stone)
            }
        }
        sum - 2 * dp[target as usize]
    }
}

#[test]
fn test() {
    let ret = Solution::last_stone_weight_ii(vec![2,7,4,1,8,1]);
    println!("{}", ret)
}